### ISSUE TITLE
autofix for dead_null_aware_expression

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#47159]
It would be great if `dart fix` could fix `dead_null_aware_expression`.

In particular, consider an API like this:

```dart
int? foo() => 0;

void main() {
  int x = foo() ?? 0;
}
```

If I change the `foo` method to be more precise in its return type by making it non-nullable, I get a `dead_null_aware_expression` info in the `main` function:

```dart
int foo() => 0;

void main() {
  int x = foo() ?? 0; // dead_null_aware_expression
}
```

It would be great if `dart fix` could just remove the `??` operator (and the right operand) for me.

https://github.com/flutter/flutter/pull/89451 is an example of an API change that triggered this quite a bit.

**Comment by asashour** [comment]
https://dart-review.googlesource.com/c/sdk/+/212824

**Comment by jcollins-g** [comment]
I don't think it would be safe to always remove dead null aware expressions in all codebases.  There's currently no way to specify that it is safe in a particular codebase or for a particular run (#49224).   Expanding this fix so it can fix all instances in a file (although not through `dart fix`) would help reduce developer effort in the given case, though, I'll give that a try.

**Comment by Hixie** [comment]
what's an example where it's not safe? I guess when you need to be compatible with two versions of the API, one of which _is_ nullable?

**Comment by bwilkerson** [comment]
Another example, which is currently being discussed elsewhere, is when you've run the null-safety migration tool and it's made some suboptimal choices. If we were to bulk apply this fix then the information about the code that users need to validate would be lost.

It's possible that over time, as more of the community is migrated, that the number of cases where such a fix might be detrimental will decrease to the point where we can safely enable it.

**Comment by Hixie** [comment]
It would make sense for cases like this for `dart fix --apply` to require another argument, like `--include-unsafe-fixes= dead_null_aware_expression` (maybe in the normal mode it could include a message saying something like "Detected 3 cases of dead_null_aware_expression that were not fixed because they could not be guaranteed to be correct. Use --include-unsafe-fixes to force fixes for these cases as well. See https://dart.dev/... for details." or something).

**Comment by bwilkerson** [comment]
We have definitely talked a few times about adding support for specifying a set of diagnostics to fix and/or a set of fixes to apply. We just haven't gotten to the point of starting a design doc where we can discuss the various use cases and how best to structure the UI.

**Comment by jcollins-g** [comment]
@Hixie I filed https://github.com/dart-lang/sdk/issues/49224 last week when I bumped into this, good to know we're on the same page.

**Comment by bwilkerson** [comment]
There is now a fix for this, so I'm going to close this issue.
